File: loom-tools/src/loom_tools/tokens/rotation.py
"""Rotation cursor for one-per-account distribution of concurrent dispatches.

Issue #3909: the ranked selection tier used to pick the single best-ranked
(or a random top-N) account, so a burst of N concurrent dispatches stacked all
N onto one account — exhausting its 5h limit while others sat idle. This module
provides a persistent, concurrency-safe *rotation cursor* so that consecutive
selections (whether sequential or concurrent) round-robin across the available
accounts one-per-account, in rotating order, until they wrap.

Mechanism: a monotonic integer counter is stored in ``.loom/tokens/.rotation_cursor``.
Each ``next_rotation_index`` call, under a mkdir-based lock (POSIX atomic;
``flock`` is unavailable on stock macOS — see ``_locking.MkdirLock``):

    1. reads the current counter (initializing it to a *random* offset the
       first time, so sibling daemons in different repos that share an account
       pool but not this cursor file start de-correlated rather than all
       colliding on index 0),
    2. computes ``counter % modulus`` as the index to return,
    3. writes ``counter + 1`` back.

Because concurrent callers each acquire the lock and take the next distinct
counter value, N concurrent selections over M available accounts receive N
consecutive counter values → ``min(N, M)`` distinct indices, wrapping past M.
Using a monotonic counter modulo the *current* modulus means the round-robin
adapts automatically as accounts drop in/out of the available set.

This module is import-safe: no I/O occurs at import time.
"""

from __future__ import annotations

import random
from pathlib import Path

from loom_tools.tokens._locking import MkdirLock

# Keep the initial random offset well within a range that never overflows a
# text int and stays comfortably above any realistic account count so the
# first modulo is uniformly distributed across accounts.
_INIT_OFFSET_BOUND = 1 << 30


def _cursor_path(tokens_dir: Path) -> Path:
    return tokens_dir / ".rotation_cursor"


def _lock_path(tokens_dir: Path) -> Path:
    return tokens_dir / ".rotation_cursor.lock"


def _read_counter(cursor_file: Path) -> int | None:
    """Read the stored monotonic counter, or ``None`` if missing/malformed."""
    try:
        raw = cursor_file.read_text(encoding="utf-8").strip()
    except OSError:
        return None
    if not raw:
        return None
    try:
        return int(raw)
    except ValueError:
        return None


def next_rotation_index(
    tokens_dir: Path,
    modulus: int,
    rng: random.Random,
) -> int:
    """Return the next round-robin index in ``[0, modulus)`` and advance the cursor.

    Args:
        tokens_dir: The ``.loom/tokens/`` directory holding the cursor file.
        modulus: Number of available accounts to rotate across (must be >= 1).
        rng: Random source used only to seed the cursor's initial offset the
            first time (for cross-repo de-correlation). Subsequent calls are
            deterministic round-robin regardless of ``rng``.

    The read-compute-write cycle is guarded by a mkdir-based lock so concurrent
    callers receive distinct consecutive counter values (one-per-account).

    On any lock-timeout or I/O failure this degrades gracefully to a random
    index rather than raising — selection must never hard-fail on cursor
    bookkeeping while accounts remain available (preserves the #3907 floor).
    """
    if modulus <= 1:
        # A single (or empty) window has nothing to rotate across.
        return 0

    cursor_file = _cursor_path(tokens_dir)
    try:
        with MkdirLock(_lock_path(tokens_dir)):
            counter = _read_counter(cursor_file)
            if counter is None:
                # First use (or a corrupt cursor): start at a random offset so
                # sibling spawners that don't share this file don't all begin
                # at index 0. Subsequent calls advance deterministically.
                counter = rng.randrange(_INIT_OFFSET_BOUND)
            index = counter % modulus
            try:
                cursor_file.write_text(str(counter + 1), encoding="utf-8")
            except OSError:
                # Best-effort persistence; still return a valid index.
                pass
            return index
    except (TimeoutError, OSError):
        # Never wedge selection on cursor bookkeeping — spread randomly.
        return rng.randrange(modulus)

File: loom-daemon/src/tokens_pool/rotation.rs
//! Rotation cursor for one-per-account distribution of concurrent dispatches,
//! ported from `loom_tools.tokens.rotation` (issue #3909).
//!
//! A monotonic counter persisted at `.loom/tokens/.rotation_cursor` under a
//! [`super::locking::MkdirLock`]. Each call reads the counter (initializing
//! to a random offset the first time so sibling repos sharing an account
//! pool but not this cursor file de-correlate), computes `counter % modulus`,
//! and writes `counter + 1` back — see the Python docstring for the full
//! rationale.

use std::path::Path;

use super::locking::MkdirLock;
use super::rng::Rng;

/// Keep the initial random offset comfortably above any realistic account
/// count so the first modulo is uniformly distributed.
const INIT_OFFSET_BOUND: u64 = 1 << 30;

fn cursor_path(tokens_dir: &Path) -> std::path::PathBuf {
    tokens_dir.join(".rotation_cursor")
}

fn lock_path(tokens_dir: &Path) -> std::path::PathBuf {
    tokens_dir.join(".rotation_cursor.lock")
}

fn read_counter(cursor_file: &Path) -> Option<u64> {
    let raw = std::fs::read_to_string(cursor_file).ok()?;
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return None;
    }
    trimmed.parse::<u64>().ok()
}

/// Return the next round-robin index in `[0, modulus)` and advance the
/// cursor. `modulus <= 1` returns `0` without touching the cursor file (a
/// single/empty window has nothing to rotate across).
///
/// On any lock-timeout or I/O failure this degrades gracefully to a random
/// index rather than propagating an error — selection must never hard-fail
/// on cursor bookkeeping while accounts remain available.
pub fn next_rotation_index(tokens_dir: &Path, modulus: usize, rng: &mut Rng) -> usize {
    if modulus <= 1 {
        return 0;
    }

    let cursor_file = cursor_path(tokens_dir);
    match MkdirLock::acquire(&lock_path(tokens_dir)) {
        Ok(_lock) => {
            let counter = read_counter(&cursor_file)
                .unwrap_or_else(|| rng.gen_range(INIT_OFFSET_BOUND as usize) as u64);
            let index = (counter % modulus as u64) as usize;
            let _ = std::fs::write(&cursor_file, (counter + 1).to_string());
            index
        }
        Err(_) => rng.gen_range(modulus),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn modulus_one_returns_zero_without_cursor_file() {
        let tmp = tempfile::tempdir().unwrap();
        let mut rng = Rng::seeded(1);
        assert_eq!(next_rotation_index(tmp.path(), 1, &mut rng), 0);
        assert!(!cursor_path(tmp.path()).exists());
    }

    #[test]
    fn modulus_zero_returns_zero() {
        let tmp = tempfile::tempdir().unwrap();
        let mut rng = Rng::seeded(1);
        assert_eq!(next_rotation_index(tmp.path(), 0, &mut rng), 0);
    }

    #[test]
    fn rotates_one_per_account_across_calls() {
        let tmp = tempfile::tempdir().unwrap();
        let mut rng = Rng::seeded(5);
        // Pre-seed the cursor at a known value so the sequence is
        // deterministic and doesn't depend on the random initial offset.
        std::fs::write(cursor_path(tmp.path()), "0").unwrap();
        let modulus = 3;
        let indices: Vec<usize> = (0..7)
            .map(|_| next_rotation_index(tmp.path(), modulus, &mut rng))
            .collect();
        assert_eq!(indices, vec![0, 1, 2, 0, 1, 2, 0]);
    }

    #[test]
    fn adapts_when_modulus_changes() {
        let tmp = tempfile::tempdir().unwrap();
        let mut rng = Rng::seeded(5);
        std::fs::write(cursor_path(tmp.path()), "4").unwrap();
        // counter=4, modulus=3 -> index 1, counter becomes 5.
        assert_eq!(next_rotation_index(tmp.path(), 3, &mut rng), 1);
        // counter=5, modulus=2 -> index 1.
        assert_eq!(next_rotation_index(tmp.path(), 2, &mut rng), 1);
    }

    #[test]
    fn malformed_cursor_falls_back_to_random_seed() {
        let tmp = tempfile::tempdir().unwrap();
        std::fs::write(cursor_path(tmp.path()), "not-a-number").unwrap();
        let mut rng = Rng::seeded(1);
        let idx = next_rotation_index(tmp.path(), 4, &mut rng);
        assert!(idx < 4);
        // The cursor should have been rewritten to a valid integer.
        let contents = std::fs::read_to_string(cursor_path(tmp.path())).unwrap();
        assert!(contents.trim().parse::<u64>().is_ok());
    }
}
